### DISCUSSION TITLE
Error expected struct `IsoSqlDefaultKeyword`, found struct `DoesNotSupportDefaultKeyword`

### DISCUSSION THREAD

**Discussion by hasezoey** [diesel-rs/diesel#3115] [Q&A] [answered]
### Versions

- **Rust:** 1.59.0
- **Diesel:** 2.0.0 (f66844da50febfd3953a815f31cb330d7ef71885)
- **Database:** sqlite
- **Operating System** linux (manjaro 21.2.5)

### Feature Flags

- **diesel:** sqlite

## Problem Description
I am getting a (very long) Error when trying to do a `.execute` on a INSERT + ON CONFLICT + SET

### What are you trying to accomplish?
Trying to do a INSERT with ON CONFLICT and SET (and then execute)

### What is the expected output?
no error

### What is the actual output?
error

### Are you seeing any additional errors?

<details>
<summary>Full Error</summary>

```txt
error[E0277]: the trait bound `diesel::query_builder::insert_statement::batch_insert::BatchInsert<Vec<diesel::query_builder::insert_statement::ValuesClause<(DefaultableColumnInsertValue<ColumnInsertValue<columns::media_id, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::provider, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::title, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>), table>>, table, (), false>: QueryFragment<Sqlite, sqlite::backend::SqliteBatchInsert>` is not satisfied
    --> src/lib.rs:28:14
     |
28   |             .execute(&mut connection).unwrap();
     |              ^^^^^^^ the trait `QueryFragment<Sqlite, sqlite::backend::SqliteBatchInsert>` is not implemented for `diesel::query_builder::insert_statement::batch_insert::BatchInsert<Vec<diesel::query_builder::insert_statement::ValuesClause<(DefaultableColumnInsertValue<ColumnInsertValue<columns::media_id, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::provider, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::title, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>), table>>, table, (), false>`
     |
     = help: the following implementations were found:
               <diesel::query_builder::insert_statement::batch_insert::BatchInsert<V, Tab, QId, HAS_STATIC_QUERY_ID> as QueryFragment<DB>>
               <diesel::query_builder::insert_statement::batch_insert::BatchInsert<Vec<diesel::query_builder::insert_statement::ValuesClause<V, Tab>>, Tab, QId, HAS_STATIC_QUERY_ID> as QueryFragment<DB, PostgresLikeBatchInsertSupport>>
     = note: required because of the requirements on the impl of `QueryFragment<Sqlite>` for `diesel::query_builder::insert_statement::batch_insert::BatchInsert<Vec<diesel::query_builder::insert_statement::ValuesClause<(DefaultableColumnInsertValue<ColumnInsertValue<columns::media_id, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::provider, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::title, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>), table>>, table, (), false>`
     = note: 2 redundant requirements hidden
     = note: required because of the requirements on the impl of `QueryFragment<Sqlite>` for `InsertStatement<table, diesel::query_builder::upsert::on_conflict_clause::OnConflictValues<diesel::query_builder::insert_statement::batch_insert::BatchInsert<Vec<diesel::query_builder::insert_statement::ValuesClause<(DefaultableColumnInsertValue<ColumnInsertValue<columns::media_id, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::provider, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::title, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>), table>>, table, (), false>, diesel::query_builder::upsert::on_conflict_target::ConflictTarget<(columns::media_id, columns::provider)>, diesel::query_builder::upsert::on_conflict_actions::DoUpdate<diesel::query_builder::update_statement::changeset::Assign<columns::title, diesel::query_builder::upsert::on_conflict_actions::Excluded<columns::title>>>>>`
     = note: required because of the requirements on the impl of `ExecuteDsl<diesel::SqliteConnection, Sqlite>` for `InsertStatement<table, diesel::query_builder::upsert::on_conflict_clause::OnConflictValues<diesel::query_builder::insert_statement::batch_insert::BatchInsert<Vec<diesel::query_builder::insert_statement::ValuesClause<(DefaultableColumnInsertValue<ColumnInsertValue<columns::media_id, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::provider, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::title, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>), table>>, table, (), false>, diesel::query_builder::upsert::on_conflict_target::ConflictTarget<(columns::media_id, columns::provider)>, diesel::query_builder::upsert::on_conflict_actions::DoUpdate<diesel::query_builder::update_statement::changeset::Assign<columns::title, diesel::query_builder::upsert::on_conflict_actions::Excluded<columns::title>>>>>`
note: required by a bound in `diesel::RunQueryDsl::execute`
    --> /home/hasezoey/.local/cargo/git/checkouts/diesel-6e3331fb3b9331ec/f66844d/diesel/src/query_dsl/mod.rs:1391:15
     |
1391 |         Self: methods::ExecuteDsl<Conn>,
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `diesel::RunQueryDsl::execute`

error[E0271]: type mismatch resolving `<Sqlite as SqlDialect>::InsertWithDefaultKeyword == IsoSqlDefaultKeyword`
    --> src/lib.rs:28:14
     |
28   |             .execute(&mut connection).unwrap();
     |              ^^^^^^^ expected struct `IsoSqlDefaultKeyword`, found struct `DoesNotSupportDefaultKeyword`
     |
     = note: required because of the requirements on the impl of `CanInsertInSingleQuery<Sqlite>` for `diesel::query_builder::insert_statement::batch_insert::BatchInsert<Vec<diesel::query_builder::insert_statement::ValuesClause<(DefaultableColumnInsertValue<ColumnInsertValue<columns::media_id, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::provider, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::title, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>), table>>, table, (), false>`
     = note: 1 redundant requirement hidden
     = note: required because of the requirements on the impl of `CanInsertInSingleQuery<Sqlite>` for `diesel::query_builder::upsert::on_conflict_clause::OnConflictValues<diesel::query_builder::insert_statement::batch_insert::BatchInsert<Vec<diesel::query_builder::insert_statement::ValuesClause<(DefaultableColumnInsertValue<ColumnInsertValue<columns::media_id, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::provider, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::title, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>), table>>, table, (), false>, diesel::query_builder::upsert::on_conflict_target::ConflictTarget<(columns::media_id, columns::provider)>, diesel::query_builder::upsert::on_conflict_actions::DoUpdate<diesel::query_builder::update_statement::changeset::Assign<columns::title, diesel::query_builder::upsert::on_conflict_actions::Excluded<columns::title>>>>`
     = note: required because of the requirements on the impl of `QueryFragment<Sqlite>` for `InsertStatement<table, diesel::query_builder::upsert::on_conflict_clause::OnConflictValues<diesel::query_builder::insert_statement::batch_insert::BatchInsert<Vec<diesel::query_builder::insert_statement::ValuesClause<(DefaultableColumnInsertValue<ColumnInsertValue<columns::media_id, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::provider, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::title, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>), table>>, table, (), false>, diesel::query_builder::upsert::on_conflict_target::ConflictTarget<(columns::media_id, columns::provider)>, diesel::query_builder::upsert::on_conflict_actions::DoUpdate<diesel::query_builder::update_statement::changeset::Assign<columns::title, diesel::query_builder::upsert::on_conflict_actions::Excluded<columns::title>>>>>`
     = note: required because of the requirements on the impl of `ExecuteDsl<diesel::SqliteConnection, Sqlite>` for `InsertStatement<table, diesel::query_builder::upsert::on_conflict_clause::OnConflictValues<diesel::query_builder::insert_statement::batch_insert::BatchInsert<Vec<diesel::query_builder::insert_statement::ValuesClause<(DefaultableColumnInsertValue<ColumnInsertValue<columns::media_id, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::provider, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::title, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Text, &String>>>), table>>, table, (), false>, diesel::query_builder::upsert::on_conflict_target::ConflictTarget<(columns::media_id, columns::provider)>, diesel::query_builder::upsert::on_conflict_actions::DoUpdate<diesel::query_builder::update_statement::changeset::Assign<columns::title, diesel::query_builder::upsert::on_conflict_actions::Excluded<columns::title>>>>>`
note: required by a bound in `diesel::RunQueryDsl::execute`
    --> /home/hasezoey/.local/cargo/git/checkouts/diesel-6e3331fb3b9331ec/f66844d/diesel/src/query_dsl/mod.rs:1391:15
     |
1391 |         Self: methods::ExecuteDsl<Conn>,
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `diesel::RunQueryDsl::execute`

Some errors have detailed explanations: E0271, E0277.
For more information about an error, try `rustc --explain E0271`.
error: could not compile `diesel_execute_error` due to 2 previous errors
```

</details>

### Steps to reproduce

1. `git clone https://github.com/hasezoey/diesel_execute_error.git`
2. `cd diesel_execute_error`
3. `cargo build --tests`
4. observe error

## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues)  and the [disussion forum](https://github.com/diesel-rs/diesel/discussions)  for similar possible closed issues.
 
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)
- [x] This issue can be reproduced without requiring a third party crate

---

PS: i am still new to SQL in general, and from what i can tell from the documentation i am not doing anything wrong
(documentation for reference is running `cargo doc` in the provided reproduction repository and navigating to `diesel/upsert/struct.IncompleteOnConflict.html#use-excluded-to-get-the-rejected-value` and `diesel/query_builder/struct.InsertStatement.html#specifying-multiple-columns-as-the-target`)

**Comment by weiznich** [comment, marked as answer]
This is expected behaviour, as the underlying backend does not support the combination of features you've chosen for your query. You can have one of the following variants:
* "Batch insert" + "default" insert values, but no upsert.
* Batch insert + upsert, but no default values
* Upsert + "default" values, but no batches.

The underlying reason is that sqlite does not support the `DEFAULT` keyword as insert value. This requires us to leave out columns that request a "default value" via `Insertable`. That in turn means we cannot do a real bulk insert there and generate a single query per insert, which is something we don't support (yet?) for upsert statements. (It might turn out that this is not even possible to support at all. Happy to see some investigation there.)

**EDIT:**
How to fix that?

There are two options: Either don't use a bulk insert, so one query per element. Or Disable the default value generation via the [corresponding attribute](https://docs.diesel.rs/master/diesel/prelude/derive.Insertable.html#optional-container-attributes)

(As a general note: Most error messages that state that some implementation is missing, related to diesel indicates that you just tried to write a query that would fail, not that diesel is broken. Questions about such error messages normally belong to the discussion forum.)

**Reply by hasezoey** [reply]
> The underlying reason is that sqlite does not support the DEFAULT keyword as insert value

where exactly is this implicitly used? (because from what i can tell, in this example i did not use `DEFAULT` anywhere, maybe `_id`?)

> (As a general note: Most error messages that state that some implementation is missing, related to diesel indicates that you just tried to write a query that would fail, not that diesel is broken. Questions about such error messages normally belong to the discussion forum.)

yes, but the error message was so long and hard to understand, so i though it may be a problem i didnt know about, also like i said i am still relatively new to SQL / sqlite, so i though to maybe report this to get a better error message in the end (if possible)

**Reply by weiznich** [reply]
> where exactly is this implicitly used? (because from what i can tell, in this example i did not use DEFAULT anywhere, maybe _id?)

That's something the `#[derive(Insertable)]` does by default. [The corresponding guide](https://diesel.rs/guides/all-about-inserts.html) has some details about this. (Note: As this guide is written for the 1.4.x release, it does not contain any details about how to disable this behaviour as this is newly added to the master branch. Anyway, it's generally to be expected that if you use a not officially released version that you need to use incomplete documentation.)

> yes, but the error message was so long and hard to understand, so i though it may be a problem i didnt know about, also like i said i am still relatively new to SQL / sqlite, so i though to maybe report this to get a better error message in the end (if possible)

It's unfortunately not really possible for us to improve the error messages, as they are emitted by rustc. If you really feel that this should be better, please fill a bug in their issue tracker. They generally try to improve the messages on their side. One concrete improvement I could see here is that rustc collapse both error messages in a single message, stating that `ExecuteDsl` is just not implemented for your query type, because that's the thing both messages are trying to tell you in a quite complicated way.

**Reply by hasezoey** [reply]
> Anyway, it's generally to be expected that if you use a not officially released version that you need to use incomplete documentation.)

i know, thanks for providing the information

> You can have one of the following variants:

i will try to find the right variant for this case, thanks for providing the insight

**Reply by hasezoey** [reply]
one last question, to clarify if i have understood correctly:
`treat_none_as_default_value` is by default `true` and needs to be explicitly disabled, and when it is `true` it changes values on diesel's side (before executing a SQL) from a `None` value to known Default values, and when `false` it changes `None` to a SQL `NULL` and so lets the SQL sever (in this case SQLite) handle it?

**Reply by hasezoey** [reply]
small update:
> There are two options: Either don't use a bulk insert, so one query per element. Or Disable the default value generation via the [corresponding attribute](https://docs.diesel.rs/master/diesel/prelude/derive.Insertable.html#optional-container-attributes)

i just tried setting `treat_none_as_default_value` to either `false` or `true` on both `Media`(`Queryable`) and `InsMedia`(`Insetable`), but the error did not change (while still doing bulk)

**Reply by weiznich** [reply]
> treat_none_as_default_value is by default true and needs to be explicitly disabled, and when it is true it changes values on diesel's side (before executing a SQL) from a None value to known Default values, and when false it changes None to a SQL NULL and so lets the SQL sever (in this case SQLite) handle it?

It's a bit different. Both are handled at database side. For `treat_none_as_default_value = true` (which is the default if you not set that explicitly) we normally emit a `DEFAULT` value. See [the postgres documentation](https://www.postgresql.org/docs/current/sql-insert.html) (The `VALUES ( { expression | DEFAULT }` part is relevant) for details. Sqlite does not support this keyword, so we emulate support by omitting the corresponding column from the insert statement. So a 
```sql
INSERT INTO users(id, name) VALUES (DEFAULT, "Jane");
```
postgresql insert statement is emulated by just doing:
```sql
INSERT INTO users(name) VALUES ("Jane");
```
at the sqlite side. That works fine for single row insert statements, but not for bulk row insert statements. There we can only emulate this behaviour on sqlite by emitting a single query per row. That's done by default if you not opt out the default value support at the derive. This support is only implemented for normal insert statements mostly because nobody has extended it for `upsert` statements yet.

Setting `treat_none_as_default_value = false` omits a `NULL` value for `None` fields. That's different than setting the default value. You could for example have column that are nullable and have a distinct default value. Setting `NULL` explicitly will set the column to `NULL`, not to the default value.

> i just tried setting treat_none_as_default_value to either false or true on both Media(Queryable) and InsMedia(Insetable), but the error did not change (while still doing bulk)

Seems like this is another case where an corresponding impl is missing to support this feature. That means the only way to do this now is to perform on upsert statement per value.

**Reply by mohe2015** [reply]
Could this be converted to an issue please? It would be great for ergonomics if upserts would also work.

**Reply by weiznich** [reply]
@mohe2015 I'm not sure what you would expect to happen if that's converted to an issue. That does not magically fix anything, especially the point that sqlite just misses support for some required SQL syntax part. With the current state I do not see any actionable possible fix here. For open issues I would expect that they are at least somewhat actionable, which is just not the case here. I'm happy to convert this to an issue or accept a PR for this if someone can suggest a solution.
